src/gui/toolbarMenubar/model/ToolbarEntry.cpp

```diff
@@ -49,6 +45,13 @@ auto ToolbarEntry::removeItemById(int id) -> bool {
 
 auto ToolbarEntry::insertItem(string item, int position) -> int {
     auto* it = new ToolbarItem(std::move(item));
+    // This is required as the position generated by gtk_toolbar_get_drop_index
+    // is 1 indexed with the notable exception that the beginning of the toolbar
+    // is still 0.
+    // Hence, the possible input positions are [0, 2, 3, 4, ...] which need to be mapped
+    // to [0, 1, 2, 3, ...] to work consistently with the the entries vector.
+    if (position >= 1)
+        position--;
```

Comment:
Dafuq? Isn't there be a more sane approach? What is the cause of this inconsistency?

Reply:
Let us assume, the implementation of gtk_toolbar_get_drop_index is correct, then we have a bug in the input values or with the handling of the returnvalue.
So we should fix that, instead of creating a fixup *after* the call.

I think the bug is in `ToolbarAdapter::toolbarDragDataReceivedCb` and `ToolbarAdapter::toolbarDragMotionCb`

For me, the following looks wrong:
```
    // x,y are already in toolbar coordinates as the gtk_toolbar_get_drop_index() specification requires.
    // However, without this translation we have an unwanted vertical offset.
```

Reply:
You're right it's not a bug in gtk --that would have been weird.

However, it took me a while to find the culprit. The `ToolbarAdapter` adds a hidden`GtkToolItem` at construction at the `0`th index in the toolbar (see https://github.com/xournalpp/xournalpp/blob/22e60aef9d2860aacefd20f27041c72c3569fb00/src/gui/dialog/toolbarCustomize/ToolbarAdapter.cpp#L111-L126)

From the context of the code it seems that this was used at some point to give the toolbar some size in case no other items were in the toolbar yet.
However, I checked and I didn't find any weird functionallity when simply removing everything apart from 

```cpp
gtk_widget_show(this->w);
```

@Febbe : Do you know if this piece of code serves any other purpose that I didn't see?

If not I would vote for removing it.

---
https://github.com/xournalpp/xournalpp/blob/22e60aef9d2860aacefd20f27041c72c3569fb00/src/gui/dialog/toolbarCustomize/ToolbarAdapter.cpp#L264-L269

Also seems weird to me but without it the position of dropped items in vertical toolbars becomes wrong.

Reply:
I think, setting the size should be done by setting the size directly, instead of inserting an dummy item.

The dummy item / setting of the size directly may be required to customize the toolbars.

The fix is nothing else than a base transformation, so it's correct, when the target base is the same as the calling widget of those signals.

Reply:
I vote for removing it. I checked the case I mentioned, and it does not make any difference.

Reply:
:sweat_smile: Wow that's old. Yeah I think that some of the css styling capabilities might have made this superfluous at some point.

I'll remove it then. 

I'll still analyze why https://github.com/xournalpp/xournalpp/blob/22e60aef9d2860aacefd20f27041c72c3569fb00/src/gui/dialog/toolbarCustomize/ToolbarAdapter.cpp#L264-L269 is required, as I would naively expect that a event handler of a toolbar already creates properly transformed coordinates. However, if I don't get anywhere with this I'll update the PR the latest by the weekend.

Reply:
I think it has something to do with the fact, that drag and drop is used at minimum across the hole application. But first I am confused, why this is not documented by GTK anywhere, it is also not required to be in AppWindow coordinates, since this callback is only triggered for the connected widget.

Reply:
I played around with this some more and I'm starting to be convinced that this is a bug in gtk :see_no_evil: 

The x coordinates are handled within the coordinate system of the toolbar, which can be seen when customizing the floating toolbar.
Thus the current version had even a bug because it translated `x` coordinates as well (my last commit fixes this).
The y coordinates, however, somehow require the AppWindow coordinate system otherwise the `pos`ition is off (see vertical toolbars).

I added a wrapper around this weird behaviour of gtk in my latest commit. Hopefully this is verbose enough :innocent:

Reply:
@Febbe : I rebased the PR yesterday and I think it should be good to go again. What do you think?

Reply:
Please add a comment directly above the line of this bug and reference it to [this](https://github.com/xournalpp/xournalpp/pull/2980#discussion_r611072211) comment of yours. This way we will know why it was done this way.